File: simple-chatbot/update_kb.py
import json
import re

def parse_faq(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Split by "Q[0-9]+:" to find questions
    # This regex looks for Q followed by digits and a colon
    pattern = re.compile(r'Q(\d+):(.*?)\nA:(.*?)(?=\nQ\d+:|$)', re.DOTALL)
    
    # The file structure is a bit loose, let's try a more robust split
    # We can split by "Q<number>:"
    
    entries = []
    
    # Normalize newlines
    content = content.replace('\r\n', '\n')
    
    # Find all Q&A blocks
    # We'll iterate through the file line by line to be safer
    lines = content.split('\n')
    
    current_q = ""
    current_a = ""
    current_id = 0
    capturing_a = False
    
    qa_pairs = []
    
    for line in lines:
        line = line.strip()
        if not line:
            continue
            
        # Check for Question
        q_match = re.match(r'^Q(\d+):\s*(.*)', line)
        if q_match:
            # Save previous if exists
            if current_q:
                qa_pairs.append({'q': current_q, 'a': current_a.strip()})
            
            current_id = q_match.group(1)
            current_q = q_match.group(2)
            current_a = ""
            capturing_a = False
            continue
            
        # Check for Answer start
        if line.startswith('A:'):
            capturing_a = True
            current_a += line[2:].strip() + " "
            continue
            
        # If capturing answer, append line
        if capturing_a:
            # Stop if we hit a "SL-Cyber-Security-Policy" line or Section header
            if "SL-Cyber-Security-Policy" in line or line.startswith("SECTION"):
                continue
            current_a += line + " "

    # Append last one
    if current_q:
        qa_pairs.append({'q': current_q, 'a': current_a.strip()})

    return qa_pairs

def generate_keywords(text):
    # Simple keyword extraction
    common_words = {'what', 'is', 'the', 'of', 'and', 'to', 'in', 'a', 'for', 'how', 'why', 'does', 'do', 'are', 'it', 'this', 'that', 'with', 'on', 'be', 'will', 'can', 'has', 'have', 'by', 'an', 'as', 'from', 'or'}
    words = re.findall(r'\w+', text.lower())
    keywords = [w for w in words if w not in common_words and len(w) > 2]
    return list(set(keywords))[:10] # Limit to 10 keywords

def update_knowledge_base(faq_path, kb_path):
    # Read existing KB
    try:
        with open(kb_path, 'r', encoding='utf-8') as f:
            kb = json.load(f)
    except:
        kb = []
        
    # Get max ID
    max_id = 0
    for item in kb:
        try:
            if int(item['id']) > max_id:
                max_id = int(item['id'])
        except:
            pass
            
    # Parse FAQ
    new_qa = parse_faq(faq_path)
    
    # Add to KB
    for item in new_qa:
        max_id += 1
        
        # Combine Q and A for content, or just use A?
        # The user wants the chatbot to answer these questions.
        # The "content" field is what the AI uses for context.
        # So we should put "Question: ... Answer: ..." in the content so the AI knows the context.
        
        content = f"{item['q']}\n{item['a']}"
        
        # Generate keywords from Question and Answer
        keywords = generate_keywords(item['q'] + " " + item['a'])
        
        new_entry = {
            "id": str(max_id),
            "topic": item['q'][:50] + "..." if len(item['q']) > 50 else item['q'],
            "content": content,
            "keywords": keywords
        }
        
        kb.append(new_entry)
        
    # Write back
    with open(kb_path, 'w', encoding='utf-8') as f:
        json.dump(kb, f, indent=4)
        
    print(f"Added {len(new_qa)} new entries to knowledge base.")

if __name__ == "__main__":
    faq_file = r"c:\Users\medvo\Desktop\GOV-AI CHATBOX\ChatFAQ\chatFAQ"
    kb_file = r"c:\Users\medvo\Desktop\GOV-AI CHATBOX\simple-chatbot\data\knowledge_base.json"
    update_knowledge_base(faq_file, kb_file)

File: simple-chatbot/update_kb.js

const fs = require('fs');
const path = require('path');

const RAW_FILE_PATH = 'C:/Users/medvo/Desktop/GOV-AI CHATBOX/ChatFAQ/chatFAQ';
const KB_PATH = 'C:/Users/medvo/Desktop/GOV-AI CHATBOX/simple-chatbot/data/knowledge_base.json';

// Helper to extract keywords (naive)
function extractKeywords(text) {
    const stopWords = new Set(['what', 'is', 'the', 'how', 'why', 'are', 'in', 'of', 'and', 'to', 'a', 'for', 'on', 'with', 'does', 'do', 'can', 'be', 'this', 'that', 'it', 'or', 'an']);
    const words = text.toLowerCase().replace(/[^\w\s]/g, '').split(/\s+/);
    const keywords = words.filter(w => w.length > 3 && !stopWords.has(w));
    return [...new Set(keywords)].slice(0, 10); // limited to 10
}

function parseRawFile() {
    const rawContent = fs.readFileSync(RAW_FILE_PATH, 'utf8');
    const lines = rawContent.split('\n');

    const entries = [];
    let currentQ = null;
    let currentA = null;
    let currentStartLine = 0;

    // Pattern for Questions: Q1:, Q1., 1., Q:
    // Pattern for Answers: A:, Answer:

    for (let i = 0; i < lines.length; i++) {
        let line = lines[i].trim();
        if (!line) continue;

        // Skip chat transcript artifacts
        if (line.includes("ChatGPT") || line.includes("I prefer this response") || line.includes("Just tell me what you want")) continue;

        // Detect Question
        // Regex: Starts with Q followed by number and char, or Number followed by dot.
        // e.g. "Q1:", "Q1", "1.", "Q1."
        // Also capture the text.
        const qMatch = line.match(/^(Q\d+[:.]?|\d+\.)\s*(.*)/i);

        if (qMatch) {
            // If we have a pending Q&A, push it
            if (currentQ && currentA) {
                entries.push({ q: currentQ, a: currentA.trim() });
                currentQ = null;
                currentA = null;
            } else if (currentQ && !currentA) {
                // Found a new question but haven't found answer for previous? 
                // Possibly multi-line question or previous was just a header?
                // Let's assume previous Q was abandoned or part of header text.
                // OR maybe the previous 'Q' was actually the answer content? 
                // It's safer to just overwrite currentQ if A is empty.
            }

            currentQ = qMatch[2] || line; // specific text or full line
            // Check if Question text is empty, maybe on next line?
            if (!qMatch[2] && lines[i + 1]) {
                // Peek next line
                // currentQ = lines[i+1].trim(); 
                // i++;
            }
        }
        // Detect Answer
        else if (line.match(/^(A:|Answer:)/i)) {
            currentA = ""; // Initialize answer buffer
            // Check if there is text on the same line
            const aMatch = line.match(/^(A:|Answer:)\s*(.*)/i);
            if (aMatch && aMatch[2]) {
                currentA += aMatch[2] + "\n";
            }
        }
        else {
            // Content line.
            // If we are in an Answer block, append.
            if (currentA !== null) {
                currentA += line + "\n";
            } else if (currentQ !== null) {
                // If we are in a Question block (but before Answer), maybe append to Question?
                currentQ += " " + line;
            }
        }
    }

    // Push last entry
    if (currentQ && currentA) {
        entries.push({ q: currentQ, a: currentA.trim() });
    }

    return entries;
}

function updateKnowledgeBase() {
    let kb = [];
    try {
        kb = JSON.parse(fs.readFileSync(KB_PATH, 'utf8'));
    } catch (e) {
        console.log("Creating new KB file.");
    }

    // Preserve IDs 1-7 (General Info)
    const preserved = kb.filter(item => parseInt(item.id) <= 7);

    const newEntries = parseRawFile();
    console.log(`Parsed ${newEntries.length} new Q&A pairs.`);

    let nextId = 8;

    // Add new entries
    for (const entry of newEntries) {
        // Clean up text
        const question = entry.q.replace(/SL-Cyber-Security-Policy/g, '').trim();
        const answer = entry.a.replace(/SL-Cyber-Security-Policy/g, '').trim();

        if (question.length < 5 || answer.length < 5) continue; // Skip noise

        preserved.push({
            id: String(nextId++),
            topic: question.substring(0, 100) + (question.length > 100 ? "..." : ""),
            content: question + "\n" + answer,
            keywords: extractKeywords(question + " " + answer)
        });
    }

    fs.writeFileSync(KB_PATH, JSON.stringify(preserved, null, 4));
    console.log(`Updated Knowledge Base with ${preserved.length} total entries.`);
}

updateKnowledgeBase();
